src/hotspot/share/cds/filemap.cpp

```diff
@@ -2409,10 +2409,18 @@ void FileMapInfo::dealloc_heap_region() {
 }
 #endif // INCLUDE_CDS_JAVA_HEAP
 
-void FileMapInfo::unmap_regions(int regions[], int num_regions) {
+void FileMapInfo::unmap_regions(int regions[], int num_regions, ReservedSpace rs) {
   for (int r = 0; r < num_regions; r++) {
     int idx = regions[r];
-    unmap_region(idx);
+
+    // If the region is inside an active ReservedSpace, its memory and address space will be
+    // freed when the ReservedSpace is released.
+    if (!rs.is_reserved()) {
+      unmap_region(idx);
+    } else {
+      // Treat this region as if it has been unmapped
+      region_at(idx)->set_mapped_base(nullptr);
+    }
```

Comment:
Personal taste, but I tend to think that the code becomes easier to read if you remove the negation in if/else statements.
```suggestion
    // If the region is inside an active ReservedSpace, its memory and address space will be
    // freed when the ReservedSpace is released.
    if (rs.is_reserved()) {
      // Treat this region as if it has been unmapped
      region_at(idx)->set_mapped_base(nullptr);
    } else {
      unmap_region(idx);
    }
```

It would also be good to have an assert to check that the unmapped region actually is within the provided ReservedSpace.